drivers/sdhc/sam_hsmci.c

```diff
@@ -0,0 +1,521 @@
+};
+
+// timeout multiplier shift (actual value is 1 << _mul_shift[*])
+static const uint8_t _mul_shift[] = {0, 4, 7, 8, 10, 12, 16, 20};
+static const uint8_t _mul_shift_size = 8;
+
+struct sam_hsmci_config {
+	Hsmci *base;
+	const struct atmel_sam_pmc_config clock_cfg;
+	const struct pinctrl_dev_config *pincfg;
+	struct gpio_dt_spec carrier_detect;
+};
+
+struct sam_hsmci_data {
+	bool open_drain;
+	uint8_t cmd_in_progress;
+	struct k_mutex mtx;
+};
+
+static int sam_hsmci_reset(const struct device *dev)
+{
+	const struct sam_hsmci_config *config = dev->config;
+	config->base->HSMCI_CR = HSMCI_CR_SWRST;
+	return 0;
+}
+
+static int sam_hsmci_get_host_props(const struct device *dev, struct sdhc_host_props *props)
+{
+	memset(props, 0, sizeof(*props));
+	props->f_max = _HSMCI_MAX_FREQ;
+	props->f_min = _HSMCI_MIN_FREQ;
+	/* high-speed not working yet due to limitations of the SDHC sm */
+	props->host_caps.high_spd_support = false;
+	props->power_delay = 500;
+	props->is_spi = false;
+	props->max_current_330 = 4;
+
+	return 0;
+}
+
+static int sam_hsmci_set_io(const struct device *dev, struct sdhc_io *ios)
+{
+	LOG_DBG("%s(clock=%d, bus_width=%d, timing=%d, mode=%d)", __func__, ios->clock,
+		ios->bus_width, ios->timing, ios->bus_mode);
+
+	const struct sam_hsmci_config *config = dev->config;
+	struct sam_hsmci_data *data = dev->data;
+	Hsmci *hsmci = config->base;
+
+	if (ios->clock > 0) {
+		int div_val = SOC_ATMEL_SAM_MCK_FREQ_HZ / ios->clock - 2;
+		if (div_val < 0) {
+			div_val = 0;
+		}
+		if (div_val > 0x1ff) {
+			div_val = 0x1ff;
+		}
+
+		LOG_DBG("divider: %d (freq=%d)", div_val,
+			SOC_ATMEL_SAM_MCK_FREQ_HZ / (div_val + 2));
+
+		hsmci->HSMCI_MR &= ~HSMCI_MR_CLKDIV_Msk;
+		hsmci->HSMCI_MR |=
+			((div_val & 1) ? HSMCI_MR_CLKODD : 0) | HSMCI_MR_CLKDIV(div_val >> 1);
+	}
+
+	hsmci->HSMCI_SDCR &= ~HSMCI_SDCR_SDCBUS_Msk;
+
+	switch (ios->bus_width) {
+	case SDHC_BUS_WIDTH1BIT:
+		hsmci->HSMCI_SDCR = HSMCI_SDCR_SDCBUS_1;
+		break;
+	case SDHC_BUS_WIDTH4BIT:
+		hsmci->HSMCI_SDCR = HSMCI_SDCR_SDCBUS_4;
+		break;
+	case SDHC_BUS_WIDTH8BIT:
+		/* I have yet to encounter an ATMEL SAM which supports 8 lane
+		   So, for now we'll just say NO. */
+		return -ENOTSUP;
+	default:
+		return -ENOTSUP;
+	}
+
+	data->open_drain = (ios->bus_mode == SDHC_BUSMODE_OPENDRAIN);
+
+	switch (ios->timing) {
+	case SDHC_TIMING_LEGACY:
+		hsmci->HSMCI_CFG &= ~HSMCI_CFG_HSMODE;
+		break;
+	case SDHC_TIMING_HS:
+		hsmci->HSMCI_CFG |= HSMCI_CFG_HSMODE;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int sam_hsmci_init(const struct device *dev)
+{
+	const struct sam_hsmci_config *config = dev->config;
+
+	/* Connect pins to the peripheral */
+	int ret = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		LOG_ERR("pinctrl_apply_state() => %d", ret);
+		return ret;
+	}
+	/* Enable module's clock */
+	(void)clock_control_on(SAM_DT_PMC_CONTROLLER, (clock_control_subsys_t)&config->clock_cfg);
```

Comment:
Unclear where this definition is coming from, but should this not be pulled from the HSMCI's devicetree entry, via a PHANDLE? Take a look at how other drivers interact with their clock controller device pointers: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sdhc/imx_usdhc.c#L256

Reply:
Ok, so this definition comes out as `DEVICE_DT_GET(DT_NODELABEL(pmc))`. Typically the `DT_INST_CLOCKS_CTLR` macro would be used to get a device pointer to the clock controller for this device instance, but if this is the way atmel drivers are managing clocks then I am ok following the convention here.

Reply:
I'm also following convention here.